index.js

```diff
@@ -1,23 +1,115 @@
+  getMessagesTaggedWith: 'async'
 }
 
 var initialState = {}
 
 exports.init = function (ssb, config) {
   var ScuttleTag = TagHelper(ssb)
-  return ssb._flumeUse('tags', FlumeReduce(2, reduce, map, null, initialState))
+  var view = ssb._flumeUse('tags', FlumeReduce(3, tagsReduce, tagsMap, null, initialState))
 
-  function reduce (result, item) {
+  return {
+    stream: view.stream,
+    get: view.get,
+    streamTaggedMessages,
+    getTaggedMessages,
+    streamMessagesTaggedWith,
+    getMessagesTaggedWith
+  }
+
+  function getTaggedMessages (cb) {
+    view.get((err, data) => {
+      if (err) cb(err)
+      var messages = {}
+      for (var author in data) {
+        for (var tag in data[author]) {
+          for (var msg in data[author][tag]) {
+            var timestamp = data[author][tag][msg]
+            if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+          }
+        }
+      }
+      var sortedMsgIds = map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      cb(null, sortedMsgIds)
+    })
+  }
+
+  function streamTaggedMessages (opts) {
+    return pull(
+      view.stream(opts),
+      pull.map((item) => {
+        if (item.tagKey) return item
+        var messages = {}
+        for (var author in item) {
+          for (var tag in item[author]) {
+            for (var msg in item[author][tag]) {
+              var timestamp = item[author][tag][msg]
+              if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+            }
+          }
+        }
+        return map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      })
+    )
+  }
+
+  function getMessagesTaggedWith (tagId, cb) {
+    view.get((err, data) => {
+      if (err) cb(err)
+      var messages = {}
+      for (var author in data) {
+        for (var msg in data[author][tagId]) {
+          var timestamp = data[author][tagId][msg]
+          if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+        }
+      }
+      var sortedMsgIds = map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      cb(null, sortedMsgIds)
+    })
+  }
+
+  function streamMessagesTaggedWith (tagId, opts) {
+    return pull(
+      view.stream(opts),
+      pull.filter((item) => {
+        // if `tagKey` property does not exist `item` is the index
+        // which needs to be reduced
+        var live = has(item, 'tagKey')
+        return !live || item.tagKey === tagId
+      }),
+      pull.map((item) => {
+        if (item.tagKey) return item
+        var messages = {}
+        for (var author in item) {
+          for (var msg in item[author][tagId]) {
+            var timestamp = item[author][tagId][msg]
+            if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+          }
+        }
+        return map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      })
+    )
+  }
```

Comment:
There's a gotcha I don't like with flumeview-reduce - that `view.stream` is gonna blow up in your face !
Read the docs to understand why (I tried to leave a warning there), or just switch to use ssb-query with `{ live: true }`!

Reply:
I'm going to try replacing this functionality with ssb-query. Not sure I understand what you're saying about this blowing up. I tried to handle both cases for when `{ live: true }`, since `tagKey` will only exist on the root `item` if it is an incremental live update, is that the gotcha you're talking about?